### ISSUE TITLE
Encrypted IdentityFile in conjunction with ssh-agent

### ISSUE THREAD

**Issue by pjgeorg** [ronf/asyncssh#327]
I have an issue connection to a server using pubkey authentication using a private key added to ssh-agent.
On connect I only specify the host, all other options are auto detected using default values or my ssh config.

In my ssh config I have specified an IdentityFile for this server. This IdentityFile is encrypted and has been added to the ssh-agent.
This is a typically configuration (in conjunction with the ssh config option IdentitiesOnly) to restrict identities being offered during authentication.
asyncssh is not able to currently handle this kind of configuration.
It loads the identity specified by IdentityFile and assumes that this one shall be used. It tries to decode it, which obviously fails with the following error message:
`asyncssh.public_key.KeyImportError: Passphrase must be specified to import encrypted private keys`
Indeed asyncssh should first check whether this particular identity is available from the ssh-agent.

Changing IdentitesOnly to no does not fix the issue. asynssh insists on requiring a passphrase to decode the IdentityFile.

Short term fix: Ignore all encrypted IdentityFiles that can not be decoded.

**Comment by ronf** [comment]
AsyncSSH already ignores encrypted key files stored in the default location when no passphrase is provided. However, I think the reason why that's not working for you is because of the fact that you are specifying these keys explicitly via IdentityFile in .ssh/config file. When you do that, AsyncSSH behaves the same as specifying the "client_keys" argument indicating that you want to explicitly load the keys in that list of files. In that case, it expects that you'll provide a passphrase if any of those keys are encrypted, and gives you an explicit error if you don't. Ignoring files that you have explicitly told it to load doesn't seem like the right behavior to me.

If you already have the keys loaded into ssh-agent, why are you pointing at the private key file explicitly with IdentityFile? That shouldn't be necessary. If it's because you want to use IdentitiesOnly, shouldn't you be pointing at public key files instead of private key files in that case?

Unfortunately, AsyncSSH doesn't support the IdentitiesOnly config option yet, so setting or unsetting that won't make a difference. Part of the reason I haven't implemented support for IdentitiesOnly is that I don't really understand how it is supposed to work. If the goal is just to limit which ssh-agent identities are tried, I would think that the files specified in IdentityFile would need to be public key files, not private key files. If you are going to specify private key files, there's really no reason to load them into the agent. If the reason is to avoid specifying the passphrase every time, that make sense, but having only the private key file without a passphrase isn't enough to be able to determine which of the agent's keys that encrypted file corresponds to, as it can't be converted to a public key (which is what the ssh-agent expects as key identifiers).

**Comment by pjgeorg** [comment]
> If you already have the keys loaded into ssh-agent, why are you pointing at the private key file explicitly with IdentityFile? That shouldn't be necessary. If it's because you want to use IdentitiesOnly, shouldn't you be pointing at public key files instead of private key files in that case?

As far as I known both, pointing at public or private key, is allowed/supported in this case.
Pointing at the private key has the advantage of working very well in combination with the option `AddKeysToAgent`. I.e. I am being asked for the passphrase the first time I connect, for all other connections the key loaded into ssh-agent is being used.

> If you are going to specify private key files, there's really no reason to load them into the agent.

In addition to the reason given above: I do not have stored all public keys on my local machine.

> If the reason is to avoid specifying the passphrase every time, that make sense, but having only the private key file without a passphrase isn't enough to be able to determine which of the agent's keys that encrypted file corresponds to, as it can't be converted to a public key (which is what the ssh-agent expects as key identifiers).

It seems that in this case the OpenSSH client uses the private keys fingerprint to determine which of the agent's keys that encrypted private key corresponds to.
E.g. you can get the fingerprint of encrypted private keys without decryption using `ssh-keygen -lf ~/.ssh/id_rsa` and then compare to the fingerprints of private keys loaded into the ssh-agent using `ssh-add -l`.

**Comment by ronf** [comment]
If this works, I'll have to do some experimentation here to see what is actually being sent over to ssh-agent. Normally, the "handle" you pass to ssh-agent to identify which key to use is the public version of the key encoded as a binary SSH string. So, unless you have the ".pub" version of the key file available to your client, I don't quite see how IdentitiesOnly could work with encrypted private keys. You said here that you don't have the public keys on the local machine -- are you sure about that?

I've never seen ssh-agent do anything with fingerprints. As far as I know, "ssh-add" computes these fingerprints itself from the public key handles which ssh-agent returns when you request identities from it. As for "ssh-keygen -lf", try it on a system where you have only the private key file and no public key file. When I tested that here, it returned:

    id_rsa is not a key file.

When I try it with the id_rsa.pub installed along with it, though, I get back the same fingerprint as I get from "ssh-add -l", which makes sense given that the SSH agent key handle is the same as what's in the id_rsa.pub. That's why I was thinking that IdentitiesOnly might actually rely on the public key file (and only that) and ignore any encrypted private key file that might or might not be present.

I'd be happy to look into supporting IdentitiesOnly using the public key files to determine which of the SSH agent keys should be used, if that's something you would find useful. In that case, I would ignore the encrypted private key files if present, which should address the problem you mentioned here. I don't think I can make it work without the public key files, though.

**Comment by ronf** [comment]
As an experiment, I tried removing my encrypted private key file and running "ssh-keygen -lf id_rsa" with only an id_rsa.pub file present on the system, and that worked fine, once again confirm that's where it is getting the key data to use to compute the fingerprint. In fact, "ssh-keygen -lf id_rsa" and "ssh-keygen -lf id_rsa.pub" both worked, and both generated the same output, with or without the "id_rsa" encrypted private key file present.

**Comment by ronf** [comment]
I also did a test with only an unencrypted private key file and no public key file on the system, and "ssh-keygen -lf" worked fine on that. This makes sense, as private key files have enough information in them to be able to reconstruct the public key and then feed that to the fingerprinting algorithm.

**Comment by pjgeorg** [comment]
> You said here that you don't have the public keys on the local machine -- are you sure about that?

Yes. I also just tested on my machine:
```
ssh -V
OpenSSH_8.4p1, OpenSSL 1.1.1g FIPS  21 Apr 2020

ssh-add -L
Could not open a connection to your authentication agent.

ssh-keygen
Generating public/private rsa key pair.
Enter file in which to save the key (/home/***/.ssh/id_rsa):  
Enter passphrase (empty for no passphrase): 
Enter same passphrase again: 
Your identification has been saved in /home/***/.ssh/id_rsa
Your public key has been saved in /home/***/.ssh/id_rsa.pub
The key fingerprint is:
SHA256:CzPI+F5Xa62JmPzCpt/dwKKLT0X1TBAX9liVxQZ9xlQ ***@office
The key's randomart image is:
+---[RSA 3072]----+
|          +o=.oBE|
|         . * + .B|
|        .   + .o.|
|   o . .         |
|  . o + S .      |
|   .   = + o     |
|    ..o + = .    |
|   . =+* = =     |
|    +=O++ + .    |
+----[SHA256]-----+

ssh-keygen -l -f .ssh/id_rsa.pub
3072 SHA256:CzPI+F5Xa62JmPzCpt/dwKKLT0X1TBAX9liVxQZ9xlQ ***@office (RSA)

ssh-keygen -l -f .ssh/id_rsa
3072 SHA256:CzPI+F5Xa62JmPzCpt/dwKKLT0X1TBAX9liVxQZ9xlQ ***@office (RSA)

rm .ssh/id_rsa.pub 
rm: remove regular file '.ssh/id_rsa.pub'? y

ssh-keygen -l -f .ssh/id_rsa
3072 SHA256:CzPI+F5Xa62JmPzCpt/dwKKLT0X1TBAX9liVxQZ9xlQ no comment (RSA)

ssh-keygen -v -l -f .ssh/id_rsa
debug1: load private ".ssh/id_rsa": incorrect passphrase supplied to decrypt private key
3072 SHA256:CzPI+F5Xa62JmPzCpt/dwKKLT0X1TBAX9liVxQZ9xlQ no comment (RSA)
+---[RSA 3072]----+
|          +o=.oBE|
|         . * + .B|
|        .   + .o.|
|   o . .         |
|  . o + S .      |
|   .   = + o     |
|    ..o + = .    |
|   . =+* = =     |
|    +=O++ + .    |
+----[SHA256]-----+

ssh-keygen -y -f .ssh/id_rsa
Enter passphrase: 
```

Strange that it works for me, but not for you.
I tried to find further information, however according to all information the `ssh-keygen -l -f .ss/id_rsa` command shown above should fail or ask for the passphrase.

According to `man ssh-keygen`:
```
     -l      Show fingerprint of specified public key file.  For RSA and DSA keys
             ssh-keygen tries to find the matching public key file and prints its
             fingerprint.  If combined with -v, a visual ASCII art representation
             of the key is supplied with the fingerprint.
```
The matching public key file has been deleted, how can it still find it?
Just to verify no other file (that might include the public key) has been loaded, I checked with strace:
`strace -o strace.log -t ssh-keygen -v -l -f .ssh/id_rsa`
Attached strace log: [strace.log](https://github.com/ronf/asyncssh/files/5582440/strace.log)

So the question is: How is `ssh-keygen` (at least on my machine) able to get the fingerprint of an encrypted RSA key without specifying a passphrase and having no access to the public key part?

**Comment by ronf** [comment]
The man page definitely suggests that the public key needs to be available. I can't really explain how it is working for you after you delete the public key, but I tried it here with a freshly built OpenSSH 8.4p1 on Ubuntu Linux 20.04.1 LTS and I saw the exact same behavior as I did on macOS. When running "ssh-keygen -v -l -f id_rsa", I get:

```
debug1: load public "/tmp/id_rsa": No such file or directory
debug1: load private "/tmp/id_rsa": incorrect passphrase supplied to decrypt private key
/tmp/id_rsa is not a key file.
```

Trying it again with an unencrypted private key file works fine, as does trying it when the "id_rsa.pub" is present.

The man page definitely suggests the public key needs to be available in some form, and that's consistent with what I'm seeing here, and I don't see anything in the "strace" output that explains how it could be working, unless somehow the private key file is being stored in unencrypted form by the time that was run.

Did you build the ssh-keygen binary yourself, or is it from your Linux distro? If the latter, do you know if they have added any custom patches? I noticed one difference is that the comment is coming out as "no comment", rather than defaulting to user@host. Also, I noticed that your version string is claiming it is the FIPS version of OpenSSL, but I don't see how that would explain this.

I happened to have a VM of Ubuntu 21.04 here where I was debugging another issue, so I decided to give this a try there as well. Surprisingly, it worked there! What's REALLY strange is that it worked even though I'm using the ssh-keygen binary that I built on the Ubuntu 20.04 system, where it didn't work! Both systems report "OpenSSH_8.4p1, OpenSSL 1.1.1f  31 Mar 2020" for the SSH version, and I've confirmed the binaries on the two systems are identical! However, it works on 21.04 and not 20.04, and I can't think of any external dependencies that could possibly explain that.

I'm really not sure what to make of this.

On another note, I confirmed that even on 21.04, it seems to depend on using the OpenSSH private key format. If you add "-m PEM" to the command when you create the key, it fails even there. OpenSSH format is the default now on recent versions of OpenSSH, so that was being used in all of our tests here, but it's interesting that it does depend on the key format. Unfortunately, even with the OpenSSH key format, I don't see how you could recover the public key without a passphrase. Looking at the details of the key encoding, the entire block of key data is encrypted, not just the private part, so I'm still at a loss here to explain how this could work.

Now that I can reproduce this, I can dig in a bit deeper with a debugger, but there's something very mysterious going on here.

**Comment by pjgeorg** [comment]
Thanks for further looking into this!

> With the ability to load a version of the public key, the rest of this all makes sense. You could use that to match against the key handles returned by ssh-agent, and that would allow you to filter the available keys when IdentitiesOnly is set. However, for this to work with only a private key file on the system, you'd need to be using OpenSSH 8.4p1 or later, and you'd need to be using the OpenSSH private key format (which is the default for newly-generated keys on recent OpenSSH versions).

According to the OpenSSH [ReleaseNotes](https://www.openssh.com/releasenotes.html), the default for private keys was already changed in version 7.8p1.
In addition it states that this format has been supported by OpenSSH since 2014.
Commit changing the default private key file is [this](https://github.com/openssh/openssh-portable/commit/ed7bd5d93fe14c7bd90febd29b858ea985d14d45) one.
This would mean that it is sufficient to use the OpenSSH private key file format for this to work (and a non ancient version of OpenSSH).

I can confirm that all my private keys are using the OpenSSH file format. These have all been created on a machine (i.e. on one of the *server*s) running CentOS 8 with OpenSSH 8.0p1. I have not specified the key format, i.e. the default format **is** the OpenSSH key format. I also couldn't find any CentOS 8 specific patch changing the default private key file format (Patches are available [here](https://git.centos.org/rpms/openssh/blob/c8/f/SOURCES)).

My *client* machine is running Fedora 33 with OpenSSH 8.4p1 (Patches are available [here](https://src.fedoraproject.org/rpms/openssh/tree/f33)).

I'm still a bit confused, as this **should** allow me to retrieve the public key of an encrypted private key without specifying a passphrase. However the command `ssh-keygen -y -f .ssh/id_rsa` requires a passphrase anyway. Seems like not even OpenSSH has fully implemented handling these *new* OpenSSH formated private keys.

**Comment by ronf** [comment]
Yeah - the OpenSSH private key format has been around for a while now, so I agree that part probably won't be an issue. Thanks for doing the research on exactly when that showed up. I knew it was at least as far back as 2015, as I implemented support for that format in AsyncSSH 1.0.0 which came out in April of 2015.

As for the "-y" option, I can confirm that OpenSSH 8.4p1 still requires me to provide a passphrase even though all the necessary data should be available in the clear in OpenSSH-format private key. So, that code path hasn't been updated to take advantage of this the way the "-l" code path has.

Getting back to your original request, I'm thinking I can modify AsyncSSH to allow OpenSSH-format private key files to be read by import_public_key() and related functions, and perhaps also allow this to work for unencrypted private keys in other formats. Once that's available, I'm thinking I would allow the client_keys argument to accept public keys when IdentitiesOnly is set, and then use this list of keys to filter the keys read from the ssh-agent. This should avoid the error about needing a passphrase for encrypted private key files, as long as they are in OpenSSH format.

There's some trickiness here given all the different syntaxes I allow in load_keypairs(), but I'll play around with this a bit and let you know how it goes.

**Comment by ronf** [comment]
Ok - a first cut at this support is now available as commit 7f4989b in the "develop" branch.

The import_public_keys() and related functions now have support for extracting public keys from OpenSSH-format private key files, or unencrypted private key files in other formats. Rather than forcing you to mix both private and public keys in the "client_keys" argument, though, I decided to add a new "agent_identities" argument where you can provide a list of public keys or certificates to allow to be used with the SSH agent (and this can take private keys as well, in the cases mentioned here). However, there's now no confusion about whether you want to use a private key as a local key or use it with SSH agent.

For backward compatibility with OpenSSH's "IdentitiesOnly" option, I also allow you to specify public keys and certificates via "IdentityFile" like you do in OpenSSH if you set "IdentitiesOnly" in the config file. In that case, you can mix together private and public keys, but any time you provide an unencrypted private key (or encrypted one with a passphrase) AsyncSSH will always make that a local key. In other cases, it will attempt to extract the public key from it and use that as one of the SSH agent identities.

In your case where you specify an OpenSSH-format encrypted private key without a passphrase, it will properly extract the public key from it and use it as an SSH agent identity, no longer giving an error about a passphrase being needed as long as IdentitiesOnly is set to True.

**Comment by ronf** [comment]
These changes are now released in AsyncSSH 2.5.0. Please let me know if you run into any problems with them.
